Adding icap (#3470)
owncloud/docs: modules/admin_manual/pages/configuration/server/virus-scanner-support.adoc

:toc: right
:page-aliases: configuration/server/antivirus_configuration.adoc
:clamav-url: http://www.clamav.net/index.html
:icap-url: https://tools.ietf.org/html/rfc3507
:c-icap-url: https://sourceforge.net/p/c-icap/wiki/ModulesConfiguration/

== Introduction

> When sharing files, security is a key aspect. The ownCloud {anti-virus-app-url}[Anti-Virus] extension helps by protecting against malicious software like trojans or viruses. It forwards files that are being uploaded to the ownCloud server to a  malware scanning engine before they are written to the storage. When a file is recognized as malicious, it can be logged and prevented from being uploaded to the server to ensure that files in ownCloud are free of malware. More sophisticated rules may be specified as admin in the ownCloud Webinterface menu:Admin[Settings > Security].

Reviewer: ```suggestion
When sharing files, security is a key aspect. The ownCloud {oc-marketplace-url}/apps/files_antivirus[Anti-Virus] extension helps by protecting against malicious software like trojans or viruses. It forwards files that are being uploaded to the ownCloud server to a  malware scanning engine before they are written to the storage. When a file is recognized as malicious, it can be logged and prevented from being uploaded to the server to ensure that files in ownCloud are free of malware. More sophisticated rules may be specified as admin in the ownCloud Webinterface menu:Admin[Settings > Security].
```

> Out of the box, the ownCloud Anti-Virus extension works with {clamav-url}[Clam AntiVirus(ClamAV)] as the directly supported virus scanner. It detects all forms of malware including trojans, viruses and worms and scans compressed files, executables, image files, PDF, as well as many other file types. The ownCloud Anti-Virus application relies on the underlying ClamAV virus scanning engine, to which the admin points ownCloud when configuring the application. The ClamAV virus definitions need to be kept up to date in order to provide effective protection.

Reviewer: ```suggestion
Out of the box, the ownCloud Anti-Virus extension works with {clamav-url}[Clam AntiVirus (ClamAV)] as the directly supported virus scanner. It detects all forms of malware including trojans, viruses and worms and scans compressed files, executables, image files, PDF, as well as many other file types. The ownCloud Anti-Virus application relies on the underlying ClamAV virus scanning engine, to which the admin points ownCloud when configuring the application. The ClamAV virus definitions need to be kept up to date in order to provide effective protection.
```

> Starting with Anti-Virus version 1.0.0, the app also offers an ICAP integration for Enterprise installations. Admins can integrate their favorite enterprise-grade Antivirus Scanners through the open standard {icap-url}[Internet Content Adaptation Protocol (ICAP)]. With this set up, ownCloud can delegate the scanning of uploaded files to another machine, the ICAP server. The ICAP server then checks them and either greenlights them or, if malicious code is found, treats the offending file(s) as specified in the settings and notifies the ownCloud server. ownCloud can then act accordingly and reject the upload. Offloading the anti-virus scans to another dedicated server can greatly improve performance compared to running the ClamAV virus scanner on the same machine as ownCloud.

Reviewer: ```suggestion
Starting with Anti-Virus version 1.0.0, the app also offers an ICAP integration for Enterprise installations. Admins can integrate their favorite enterprise-grade Antivirus Scanners through the open standard {icap-url}[Internet Content Adaptation Protocol (ICAP)]. With this set up, ownCloud can delegate the scanning of uploaded files to another machine, the ICAP server. The ICAP server then checks them and either greenlights them or, if malicious code is found, treats the offending file(s) as specified in the settings and notifies the ownCloud server. ownCloud can then act accordingly and based on the settings made reject the upload. Offloading the anti-virus scans to another dedicated server can greatly improve performance compared to running the ClamAV virus scanner on the same machine as ownCloud.
```
We define in the sentence above the actions

:toc: right
:page-aliases: configuration/server/antivirus_configuration.adoc
:clamav-url: http://www.clamav.net/index.html
:icap-url: https://tools.ietf.org/html/rfc3507
:c-icap-url: https://sourceforge.net/p/c-icap/wiki/ModulesConfiguration/

> :docker-url: https://hub.docker.com/r/deepdiver/icap-clamav-service

Reviewer 2: this is a pure test docker image - not for productive usage - we should not have this in the admin docs

Author: Also gone.

> :anti-virus-app-url: https://marketplace.owncloud.com/apps/files_antivirus

Reviewer: pls remove this anti-virus-app-url attribute see text below ho to use it

Author: Thanks, it's gone.

> :kav-docker-url: https://box.kaspersky.com/d/c8d8577dc2494256b45e/

Reviewer 2: not sure if we are allowed to publish this url ..... I'd remove this and just reference that you need kav .... installed in any manner

Author: Ok, gone.

== Introduction

> When sharing files, security is a key aspect. The ownCloud {anti-virus-app-url}[Anti-Virus] extension helps by protecting against malicious software like trojans or viruses. It forwards files that are being uploaded to the ownCloud server to a  malware scanning engine before they are written to the storage. When a file is recognized as malicious, it can be logged and prevented from being uploaded to the server to ensure that files in ownCloud are free of malware. More sophisticated rules may be specified in 'Advanced' mode. ownCloud administrators can find the configuration options in the 'Security' settings.

Reviewer: ...More sophisticated rules may be...
pls make this a `menu:first[second > third...]` to identify where to look for this

Author: Done, thanks.

Out of the box, the ownCloud Anti-Virus extension works with {clamav-url}[Clam AntiVirus(ClamAV)] as the officially supported virus scanner. It detects all forms of malware including trojans, viruses, and worms and scans compressed files, executables, image files, PDF, as well as many other file types. The ownCloud Anti-Virus application relies on the underlying ClamAV virus scanning engine, to which the admin points ownCloud when configuring the application. The ClamAV virus definitions need to be kept up to date in order to provide effective protection.

> From version 1.0.0 on, Anti-Virus offers an ICAP integration for Enterprise installations. This means that admins can integrate their favorite enterprise-grade Antivirus Scanners through the open standard {icap-url}[Internet Content Adaptation Protocol (ICAP)]. With this set up, ownCloud can delegate the scanning of uploaded files to another machine, the ICAP server. The ICAP server then checks them and either greenlights them or, if malicious code is found, treats the offending file(s) according to its settings and notifies the ownCloud server. ownCloud can then act accordingly and reject the upload. Offloading the anti-virus scans to another dedicated server can greatly improve performance and scale compared to running the ClamAV virus scanner on the same machine as ownCloud.

Reviewer: From version 1.0.0 on...
of what?

Author: Of the Anti-Virus app. Made it clearer: Starting with Anti-Virus version 1.0.0, the app also offers an ICAP integration...

=== Installation

- Install the {oc-marketplace-url}/apps/files_antivirus[Anti-Virus app] from the ownCloud marketplace. Alternatively, use this occ command:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} market:install files_antivirus
----

- Enable the app as admin in ownCloud under menu:[Settings > Apps] in the category `Security` or with the following occ command:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} app:enable files_antivirus
----

=== General Configuration

ICAP can be configured via the Web interface as admin user or via xref:occ_command.adoc[occ commands].

==== Antivirus Configuration Panel

Log in to ownCloud as admin via the Web interface and go to menu:Admin[Settings > Security] and you'll see the Antivirus Configuration dialog.

image:configuration/server/anti-virus-panel.png[Security Settings, width:80%]

Enter the desired values:

> * `Mode`: Set to `Daemon (ICAP)`.
> 
> * `Host`: Enter the IP addess of your ICAP server.
> 
> * `Port`: Specify the port number (default 1344).
> 
> * `Stream Length`: Set the length of streams sent to the ICAP server in bytes.
> 
> * `File size limit`: If you want to limit the file size, enter the maximum value in bytes. Default is no limit (-1).
> 
> * `ICAP request service`: Select the antivirus software you want to use: avscan for ClamAV, req for Kaspersky ScanEngine or `wwreqmod` for McAfee.
> 
> * `ICAP response header holding the virus information`: Use `X-Infection-Found` for ClamAV (avscan) and `X-Virus-ID` for KAV (req). McAfee doesn't offer response headers.
> 
> * `When infected files were found during a background scan`: Specify what to do with the flagged files. Possible values: `Delete file` or `Only log`.

Reviewer: ```suggestion
`Mode`::
Set to `Daemon (ICAP)`.

`Host`::
Enter the IP addess of your ICAP server.

`Port`::
Specify the port number (default 1344).

`Stream Length`::
Set the length of streams sent to the ICAP server in bytes.

`File size limit`::
If you want to limit the file size, enter the maximum value in bytes. Default is no limit (-1).

`ICAP request service`::
Select the antivirus software you want to use: avscan for ClamAV, req for Kaspersky ScanEngine or `wwreqmod` for McAfee.

`ICAP response header holding the virus information`::
Use `X-Infection-Found` for ClamAV (avscan) and `X-Virus-ID` for KAV (req). McAfee doesn't offer response headers.

`When infected files were found during a background scan`::
Specify what to do with the flagged files. Possible values: `Delete file` or `Only log`.
```
This renders a bit better, see screenshot:
![image](https://user-images.githubusercontent.com/3321281/115008283-3b837800-9eab-11eb-8ce3-c6bb330db90a.png)

. A file size limit applies both to background jobs and to file uploads.
. After installing ClamAV and the related tools, you will have two configuration files: `/etc/freshclam.conf` and `/etc/clamd.d/scan.conf`. 
. We recommend that you enable verbose logging in both `clamd.conf` and `freshclam.conf` until you get any kinks with your ClamAV installation worked out

== ICAP

> {icap-url}[ICAP] is an open standard supported by many antivirus products. With the release of the _Anti-Virus_ app 1.0.0, other virus scanners beside ClamAV can be used via ICAP if you are running it on an ownCloud Enterprise Edition. Currently the only supported virus scanners are _Kaspersky ScanEngine_ and _McAfee Antivirus_ although far more products might simply work.

Reviewer: With the release of the _Anti-Virus_ app 1.0.0...
I would avoid to print the version, it has no value.
You also may make `Anti-Virus` a link to the marketplace. Even the final path is not set - can be fixed later.

Reviewer 2: clamav in icap is supported as well - right? at least we know this is working .....

Author: @mmattel Well, with older versions there's no ICAP, with future versions more virus scanners might be supported.

@DeepDiver1975 Good point, waiting for information from @pmaier1.

Reviewer 3: - Before version 1.0.0 there is no ICAP at all. So might be worth to point that out, yes.
- I would add "supported **and tested**"

Author: Okay, version stays, added 'tested'.

=== Installation

> - Install the Anti-Virus app 1.0.0 from the {anti-virus-app-url}[ownCloud marketplace]. Alternatively, use this occ command:

Reviewer: remove the release info
also use {oc-marketplace-url}/apps/files_antivirus[Anti-Virus app]

Author: Done.

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} market:install files_antivirus
----

- Enable the app as admin in ownCloud under Settings > Apps or with the following occ command:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} app:enable files_antivirus
----

=== General Configuration

> ICAP ist configured via occ commands. Go to your ownCloud directory, usually `/var/www/owncloud`, and enter the following occ commands:

Reviewer 2: ist -> is

Reviewer 2: and this is miss leading.

icap settings are setup just like any other antivirus settings: there is an admin ui for it.

in addition all these settings values can be set using the occ command

Author: @DeepDiver1975 More than happy to do so. However, should we point admins to the config file: /opt/kaspersky/ScanEngine/etc/kavicapd.xml and tell them to restart kavicapd after they made changes?

Reviewer 2: that is all kav and environment specific. no idea how people install kav in their environment.
kav supports various deployment mechanisms ... we don't want to document that all .... just drop that all from my pov

Author: Okay, thanks, config file is completely gone.

* To set the IP address of your anti-virus server:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus av_host --value="172.17.0.3"
----

* Specify the port of the anti-virus server:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus av_port --value="1344"
----

* Set the mode to ICAP:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus av_mode --value="icap"
----

> The setting 'icap' triggers a grace period of 24 hours if you don't have an Enterprise license but want to test ICAP.

Reviewer: maybe set as a NOTE: to render differently?

Author: Ah, yes!

* Specify what do do with the offending file:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus av_infected_action --value="delete"
----

Possible values are 'delete' and 'only_log'.

=== Run with c-icap/ClamAV

c-icap has a built in ClamAV module. For more information see: {c-icap-url}[c-icap on sourceforge]

> An out of the box docker image is available {docker-url}[here].

Reviewer 2: as above - plz drop my docker container from the docs

Author: Removed and the kaspersky url as well.

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} occ config:app:set files_antivirus av_request_service --value="avscan"
----

Then set the response header:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus av_response_header --value="X-Infection-Found"
----

### Run with Kaspersky Anti-Virus (KAV)

Kaspersky provides {kav-docker-url}[docker images] as well.
Follow the instructions in Kaspersky ScanEngine for Kubernetes.7z.

Additional configuration: Set <Configuration><ICAPSettings><Allow204>1</Allow204></ICAPSettings></Configuration> in kavicapd.xml

Set KAV as virus scanner with the value 'req':

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus av_request_service --value="req"
----

And the respective response header:

[source,console,subs="attributes+"]
----
{occ-command-example-prefix} config:app:set files_antivirus av_response_header --value="X-Virus-ID"
----

NOTE: The older versions of KAV did not send back the virus/infection name in an ICAP header.

> In v2.0.0 the header to transport the virus can be configured. Default: No header is sent.

Reviewer: what is v2.0.0 referring to?

Author: Deleted the para break and added KAV again after the version.

> ```

Reviewer: ```
[source,console]
----
sed ...
----
```
What is ths command doing, from where should this command be run, do you need previleges (sudo)?

Reviewer 2: I'd just drop that - one should not use sed to maniplate xml ..... furthermore: people need to edit their config in what ever way fits their envrionment.

Author: Done.
